tests/karate/src/test/java/fe/net/MockBackend.java

```diff
@@ -0,0 +1,102 @@
+package fe.net;
+
+import com.intuit.karate.Logger;
+import com.intuit.karate.http.WebSocketServerBase;
+
+import common.net.MessageBuffer;
+import common.net.MessageQueue;
+import karate.io.netty.channel.Channel;
+import karate.io.netty.channel.ChannelHandlerContext;
+import karate.io.netty.channel.SimpleChannelInboundHandler;
+import karate.io.netty.handler.codec.http.websocketx.TextWebSocketFrame;
+
+import java.util.concurrent.*;
+import java.util.function.Function;
+
+/** Defines a mock backend server that is fully customisable. */
+public class MockBackend extends SimpleChannelInboundHandler<TextWebSocketFrame> {
+
+  private final Logger logger = new Logger(getClass().getSimpleName());
+
+  private final MessageQueue queue;
+  private final WebSocketServerBase server;
+  // Will be set to true once the connection is established
+  private final CompletableFuture<Boolean> connected = new CompletableFuture<>();
+
+  // Defines the rule to apply on incoming messages to produce its reply.
+  // Can be null if no reply should be sent back.
+  private Function<String, String> replyProducer = ReplyMethods.ALWAYS_VALID;
+  private Channel channel;
+
+  public MockBackend(MessageQueue queue, int port) {
+    this.queue = queue;
+    server = new WebSocketServerBase(port, "/", this);
+    logger.info("Mock Backend started");
+  }
+
+  /**
+   * Sets the reply producer of the backend.
+   *
+   * <p>It can be set to null if no reply should be sent back
+   *
+   * @param replyProducer to set
+   */
+  public void setReplyProducer(Function<String, String> replyProducer) {
+    this.replyProducer = replyProducer;
+  }
+
+  @Override
+  public void channelActive(ChannelHandlerContext ctx) {
+    channel = ctx.channel();
+    connected.complete(true);
+    logger.trace("Client connected from the server side");
+  }
+
+  @Override
+  protected void channelRead0(
+      ChannelHandlerContext channelHandlerContext, TextWebSocketFrame frame) {
+    logger.info("message received : {}", frame.text());
+    queue.onNewMsg(frame.text());
+
+    // Send back the reply
+    if (replyProducer != null) send(replyProducer.apply(frame.text()));
```

Comment:
So following above comment here the replyProducer would be basically set back to null

Reply:
I don't agree.

If we want to do a message by message response, we can use backend.send() directly inside the test.

It will be easier to understand.

Reply:
Ok